src/StackExchange.Redis/ConnectionMultiplexer.cs

```diff
@@ -2214,30 +2350,61 @@ public ConnectionMultiplexer GetSentinelMasterConnection(ConfigurationOptions co
                     return sentinelConnectionChildren[config.ServiceName];
             }
 
-            // Get an initial endpoint - try twice
-            EndPoint newMasterEndPoint = GetConfiguredMasterForService(config.ServiceName)
-                                      ?? GetConfiguredMasterForService(config.ServiceName);
+            int attempts = 0;
+            bool success = false;
+            ConnectionMultiplexer connection = null;
 
-            if (newMasterEndPoint == null)
+            do
             {
-                throw new RedisConnectionException(ConnectionFailureType.UnableToConnect,
-                    $"Sentinel: Failed connecting to configured master for service: {config.ServiceName}");
-            }
+                attempts++;
 
-            // Replace the master endpoint, if we found another one
-            // If not, assume the last state is the best we have and minimize the race
-            if (config.EndPoints.Count == 1)
-            {
-                config.EndPoints[0] = newMasterEndPoint;
-            }
-            else
+                // Get an initial endpoint - try twice
+                EndPoint newMasterEndPoint = GetConfiguredMasterForService(config.ServiceName)
+                                             ?? GetConfiguredMasterForService(config.ServiceName);
+
+                if (newMasterEndPoint == null)
+                {
+                    throw new RedisConnectionException(ConnectionFailureType.UnableToConnect,
+                        $"Sentinel: Failed connecting to configured master for service: {config.ServiceName}");
+                }
+
+                EndPoint[] slaveEndPoints = GetSlavesForService(config.ServiceName)
+                                            ?? GetSlavesForService(config.ServiceName);
+
+                // Replace the master endpoint, if we found another one
+                // If not, assume the last state is the best we have and minimize the race
+                if (config.EndPoints.Count == 1)
+                {
+                    config.EndPoints[0] = newMasterEndPoint;
+                }
+                else
+                {
+                    config.EndPoints.Clear();
+                    config.EndPoints.TryAdd(newMasterEndPoint);
+                }
+
+                foreach (var slaveEndPoint in slaveEndPoints)
+                    config.EndPoints.TryAdd(slaveEndPoint);
+
+                connection = ConnectImpl(config, log);
+
+                // verify role is master according to:
+                // https://redis.io/topics/sentinel-clients
+                if (connection.GetServer(newMasterEndPoint).Role() == RedisLiterals.master)
+                {
+                    success = true;
+                    break;
+                }
+
+                Thread.Sleep(100);
```

Comment:
Having these sleeps in the multiplexer causes all sorts of issues on resumption we found in the first iteration of Sentinel (later removed in #1403), what's the goal here with them?

Reply:
I was just trying to follow the sentinel client best practices. It says to check role for master before doing anything else and to retry finding the master from sentinel and try again if it’s not master yet.

Reply:
@mgravell thoughts?